src/main/java/org/kitteh/irc/client/library/implementation/ActorProvider.java

```diff
@@ -674,6 +679,140 @@ public String toString() {
+            this.remoteAddress = remoteAddress;
+        }
+
+        void setConnected(boolean connected) {
+            this.connected = connected;
+        }
+
+        void setNettyChannel(io.netty.channel.Channel nettyChannel) {
+            this.nettyChannel = nettyChannel;
+        }
+
+        @Override
+        IRCDCCExchangeSnapshot snapshot() {
+            throw new UnsupportedOperationException("Cannot snapshot only the exchange");
+        }
+    }
+    
+    class IRCDCCChat extends IRCDCCExchange {
+        private IRCDCCChat(@Nonnull String name) {
+            super("CHAT", name);
+        }
+
+        @Override
+        void onMessage(String msg) {
+            ActorProvider.this.client.getEventManager().callEvent(new DCCMessageEvent(ActorProvider.this.client, Collections.emptyList(), snapshot(), msg));
+        }
+
+        @Override
+        void sendCTCP() {
+            InetSocketAddress addr = (InetSocketAddress) this.getLocalAddress();
+            String ctcp = String.format("DCC CHAT chat %s %s", addr.getAddress().getHostAddress(), addr.getPort());
+            ActorProvider.this.client.sendCTCPMessage(getName(), ctcp);
+        }
+
+        @Override
+        IRCDCCChatSnapshot snapshot() {
+            return new IRCDCCChatSnapshot(this);
+        }
+    }
+
+    abstract class IRCDCCExchangeSnapshot extends IRCActorSnapshot implements DCCExchange {
+        final io.netty.channel.Channel nettyChannel;
+        private final SocketAddress localAddress;
+        private final SocketAddress remoteAddress;
+        private final boolean connected;
+        
+        private IRCDCCExchangeSnapshot(@Nonnull IRCDCCExchange actor) {
+            super(actor);
+            this.nettyChannel = actor.nettyChannel;
+            this.localAddress = actor.localAddress;
+            this.remoteAddress = actor.remoteAddress;
+            this.connected = actor.connected;
+        }
+
+        @Override
+        public SocketAddress getLocalSocketAddress() {
+            return this.localAddress;
+        }
+
+        @Override
+        public SocketAddress getRemoteSocketAddress() {
+            return this.remoteAddress;
+        }
+
+        @Override
+        public boolean isConnected() {
+            return this.connected;
+        }
+
+        @Override
+        @Nonnull
+        public String toString() {
+            return new ToStringer(this).add("client", this.getClient()).add("name", this.getName()).add("localSocketAddress", this.localAddress).add("remoteSocketAddress", this.remoteAddress).add("connected", this.connected).toString();
+        }
+    }
+    
+    class IRCDCCChatSnapshot extends IRCDCCExchangeSnapshot implements DCCChat {
+        private IRCDCCChatSnapshot(@Nonnull IRCDCCChat actor) {
+            super(actor);
+        }
+
+        @Override
+        public void sendMessage(@Nonnull String message) {
+            Sanity.nullCheck(message, "message cannot be null");
```

Comment:
Should most likely ensure the contents of the message are valid? Not sure if DCC allows different content and does not need validation.